wasm-calc11/test/static/typecheck.calc

```diff
@@ -0,0 +1,83 @@
+type Result<e,a>
+  = Left(e)
+  | Right(a)
+
+type Type
+  = TBoolean
+  | TInt
+
+type Expr<ann>
+  = EBoolean(ann, Boolean)
+  | EIf(ann, Expr(ann), Expr(ann), Expr(ann))
+  | EInt(ann, Int32)
+
+type Unit
+  = Unit
+
+type TypeError
+  = ExpectedBooleanGotInt
+  | InternalNonsense
+  | TypeMismatch(Type, Type)
+
+function outerAnnotation<ann>(expr: Expr(ann)) -> ann { 
+  case expr {
+    EInt(ann, _) -> ann, 
+    EBoolean(ann, _) -> ann, 
+    EIf(ann, _, _, _) -> ann 
+  }
+}
+
+function isBool(expr: Expr(Type)) -> Boolean { 
+  case outerAnnotation(expr) {
+    TBoolean -> True, _ -> False 
+  }
+}
+
+function typeEquals(typeA: Type, typeB: Type) -> Boolean { 
+  case (typeA, typeB) {
+    (TInt, TInt) -> True, 
+    (TBoolean, TBoolean) -> True, 
+    _ -> False 
+  }
+}
+
+function typecheck(expr: Expr(Unit)) -> Result(
+  TypeError, Expr(Type)
+) { 
+  case expr {
+    EInt(_, i) -> Right(EInt(TInt, i)), 
+    EBoolean(_, b) -> Right(EBoolean(TBoolean, b)), 
+    EIf(_, predExpr, thenExpr, elseExpr) -> {
+      case typecheck(predExpr) {
+        Left(e) -> Left(e), 
+        Right(typedPred) -> {
+          if isBool(typedPred) then
+            case (
+              typecheck(thenExpr), typecheck(elseExpr)
+            ) {
+              (Right(typedThen), Right(typedElse)) -> {
+                let tyThen = outerAnnotation(typedThen); 
+                let tyElse = outerAnnotation(typedElse); 
+                if typeEquals(tyThen, tyElse) then
+                  Right(
+                    EIf(
+                      tyThen, 
+                      typedPred, 
+                      typedThen, 
+                      typedElse
+                    )
+                  )
+                else
+                  Left(TypeMismatch(tyThen, tyElse))
+              }, 
+              (_, Left(e)) -> Left(e), 
+              (Left(e), _) -> Left(e), 
+              _ -> Left(InternalNonsense) 
```

Comment:
This is disappointing, pattern matching thought it was non-exhaustive so had to add this unreachable thing. Need to fix this.